src/Artsy/SystemContext.tsx

```diff
@@ -57,12 +59,20 @@ export const ContextProvider: SFC<ContextProps<any>> = ({
   const relayEnvironment =
     props.relayEnvironment || createRelaySSREnvironment({ user: _user })
 
+  // Listen for changes to relay data fetching state and update consumers
+  const [isFetchingData, toggleFetching] = useState(false)
+
   const providerValues = {
     ...props,
     user: _user,
     relayEnvironment,
+    isFetchingData,
   }
 
+  // Attach a loading state toggle to the relay environment which is triggered
+  // from within the middleware when fetches begin / complete.
+  relayEnvironment.toggleFetching = toggleFetching
```

Comment:
This is certainly the most direct way to get this into the middleware, but wondering what you think @alloy? We have many different ways to create a network layer across a few different pathways, but in the end everything ends up here.

Reply:
Hmm. Yeah, I can see how this makes sense. It’s not the cleanest to extend upstream objects like this, but also not unheard of. I wonder mostly, though, how are we using this? Is that in another PR, because I can’t find any usage in this PR.

If we’re not using it right now then I’d prefer we don’t add this right now either. It’s good to know _how_ we’d easily solve it if the need arises, though 👍

Reply:
Cool, that seems reasonable. Initially it was placed to wrestle a bit more control over the spinner experience but that turned out to be unnecessary because the `<Route>` component sufficiently handles it. I can remove.